src/language_server/custom_requests/ShowAst.ml

```diff
@@ -0,0 +1,72 @@
+(* Brandon Wu
+ *
+ * Copyright (C) 2019-2023 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Lsp
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* Custom LSP request for a command which will show the Semgrep
+   Generic AST of the given file.
+
+   This file largely follows the structure of ocaml-lsp, such as:
+   https://github.com/ocaml/ocaml-lsp/blob/master/ocaml-lsp-server/src/custom_requests/req_wrapping_ast_node.ml
+*)
+
+(*****************************************************************************)
+(* Request parameters *)
+(*****************************************************************************)
+
+module Request_params = struct
+  type t = { document_uri : Uri.t; named : bool }
+
+  let _params_schema =
+    `Assoc [ ("uri", `String "<DocumentUri>"); ("named", `Bool false) ]
+
+  let of_jsonrpc_params params : t option =
+    match params with
+    | Some (`Assoc [ ("uri", uri); ("named", `Bool named) ]) ->
+        let document_uri = Uri.t_of_yojson uri in
+        Some { document_uri; named }
+    | __else__ -> None
+
+  let of_jsonrpc_params_exn params : t =
+    match of_jsonrpc_params params with
+    | None -> failwith "expected jsonrpc schema matching showast"
+    | Some res -> res
+end
+
+(*****************************************************************************)
+(* Main logic *)
+(*****************************************************************************)
+
+let meth = "semgrep/showAst"
+
+let on_request (params : Jsonrpc.Structured.t option) =
+  let { Request_params.document_uri; named } =
+    Request_params.of_jsonrpc_params_exn params
+  in
+  let file = Uri.to_path document_uri in
+  let lang = Lang.lang_of_filename_exn (Fpath.v file) in
+  (* copied from -dump_ast *)
+  let { Parsing_result2.ast; _ } =
+    if named then Parse_target.parse_and_resolve_name lang file
+    else Parse_target.just_parse_with_lang lang file
+  in
+  let v = Meta_AST.vof_any (AST_generic.Pr ast) in
+  (* 80 columns is too little *)
+  Format.set_margin 120;
+  let s = OCaml.string_of_v v in
```

Comment:
Mentioned this in the VSCode PR, but depending on how you want to display the AST, you may want to turn this into some sort of JSON, instead of formatted text, so we can display it better in VSCode

Reply:
I'm not actually sure if there's a way to generate a `Yojson` value from our `AST_generic`. For now, can we move with the formatted string, and then fix it later if we see a need to make it nicer?

Reply:
I think you should be able to do
`ast |> AST_generic_to_v1.program |> Ast_generic_v1_j.string_of_program`
or another similar Ast_generic_v1_j method. There's an atd spec of the ast so I think there should be a way if that's not it.

I'm ok with the text view tho, mostly leaving this comment for if we want to come back to it in the future